grouper/permissions.py

```diff
@@ -59,6 +60,31 @@ def grant_permission(session, group_id, permission_id, argument=''):
     session.commit()
 
 
+def grant_permission_to_tag(session, tag_id, permission_id, argument=''):
+    # type: Session, int, int, str -> None
+    """
+    Grant a permission to this tag. This will fail if the (permission, argument) has already
+    been granted to this tag.
+
+    Args:
+        session(models.base.session.Sessioan): database session
+        tag_id(int): the id of the tag we're granting the permission to
+        permission_id(int): the id of the permission to be granted
+        argument(str): must match constants.ARGUMENT_VALIDATION
+
+    Throws:
+        AssertError if argument does not match ARGUMENT_VALIDATION regex
+    """
+    assert re.match(ARGUMENT_VALIDATION, argument), 'Permission argument does not match regex.'
+
+    mapping = TagPermissionMap(permission_id=permission_id, tag_id=tag_id, argument=argument)
+    mapping.add(session)
+
+    Counter.incr(session, "updates")
+
+    session.commit()
```

Comment:
consider catching the `IntegrityError` and handling the rollback here instead of the handler. keeps to direct database code confined to this module.